src/plugins/NativeTokenLimitPlugin.sol

```diff
@@ -0,0 +1,152 @@
+
+/// @title Native Token Limit Plugin
+/// @author ERC-6900 Authors
+/// @notice This plugin supports a single total native token spend limit.
+/// It tracks a total spend limit across UserOperation gas limits and native token transfers.
+/// If a paymaster is used, UO gas would not cause the limit to decrease.
+contract NativeTokenLimitPlugin is BasePlugin, IExecutionHook, IValidationHook {
+    using UserOperationLib for PackedUserOperation;
+    using EnumerableSet for EnumerableSet.Bytes32Set;
+
+    string public constant NAME = "Native Token Limit";
+    string public constant VERSION = "1.0.0";
+    string public constant AUTHOR = "ERC-6900 Authors";
+
+    mapping(address account => uint256[] limits) public limits;
+
+    error ExceededNativeTokenLimit();
+    error ExceededNumberOfEntities();
+
+    function updateLimits(uint8 functionId, uint256 newLimit) external {
+        limits[msg.sender][functionId] = newLimit;
+    }
+
+    /// @inheritdoc IExecutionHook
+    function preExecutionHook(uint8 functionId, bytes calldata data) external override returns (bytes memory) {
+        bytes calldata callData;
+        bytes4 execSelector;
+
+        // TODO: plugins should never have to do these gymnastics
+        execSelector = bytes4(data[52:56]);
+        if (execSelector == IAccountExecute.executeUserOp.selector) {
+            callData = data[56:];
+            execSelector = bytes4(callData);
+        } else {
+            callData = data[52:];
+        }
+
+        uint256 value;
+        // Get value being sent
+        if (execSelector == IStandardExecutor.execute.selector) {
+            value = uint256(bytes32(callData[36:68]));
+        } else if (execSelector == IStandardExecutor.executeBatch.selector) {
+            Call[] memory calls = abi.decode(callData[4:], (Call[]));
+            for (uint256 i = 0; i < calls.length; i++) {
+                value += calls[i].value;
+            }
+        }
+
+        uint256 limit = limits[msg.sender][functionId];
+        if (value > limit) {
+            revert ExceededNativeTokenLimit();
+        }
+        limits[msg.sender][functionId] = limit - value;
+
+        return "";
+    }
+
+    /// @inheritdoc IExecutionHook
+    function postExecutionHook(uint8, bytes calldata) external pure override {
+        revert NotImplemented();
+    }
+
+    // No implementation, no revert
+    // Runtime spends no account gas, and we check native token spend limits in exec hooks
+    function preRuntimeValidationHook(uint8 functionId, address, uint256, bytes calldata) external pure override {
+        // silence warnings
+        (functionId);
+    }
+
+    /// @inheritdoc IValidationHook
+    function preUserOpValidationHook(uint8 functionId, PackedUserOperation calldata userOp, bytes32)
+        external
+        returns (uint256)
+    {
+        // Decrease limit only if no paymaster is used
+        if (userOp.paymasterAndData.length == 0) {
```

Comment:
You could have a paymaster that fronts the gas fee and then pays it back, so technically, the UserOp has used/paid for that gas. How would this be handled?

Reply:
Interesting use case, I think this would be something like an ERC20 paymaster, pulls ERC20 tokens to pay for gas. One way to handle it is for the owner to set up a list of paymasters that, when used, would still decrement the allocated spend for that entity. This list of paymasters would be exactly the list of paymasters that have been given permission to pull funds from the account

Reply:
Since thats a really common use case, will make the change to support this. Great find!